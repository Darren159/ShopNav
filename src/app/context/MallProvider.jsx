import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import PropTypes from "prop-types";
import fetchMalls from "../../services/fetchMalls";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const initializeMalls = async () => {
      try {
        const mallList = await fetchMalls();
        setMalls(mallList);
        setCurrentMall(mallList[0]); // Set the first mall as the current mall
      } catch (error) {
        Alert.alert("Error", error.message, [{ text: "OK" }], {
          cancelable: false,
        });
      }
    };
    initializeMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
