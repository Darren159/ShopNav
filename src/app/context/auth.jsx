import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import {
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import PropTypes from "prop-types";

import { auth } from "../../../firebaseConfig";

export const AuthContext = createContext({});

export function AuthProvider({ children }) {
  const [user, setUser] = useState(null);

  useEffect(() => {
    onAuthStateChanged(auth, (userState) => {
      if (userState) {
        setUser(userState);
      } else {
        setUser(null);
      }
    });
  }, []);

  const authContextValue = useMemo(
    () => ({
      user,
      setUser,
      signin: async (email, password) => {
        try {
          await signInWithEmailAndPassword(auth, email, password);
        } catch (error) {
          let errorMessage;
          
          switch (error.code) {
            case 'auth/user-not-found':
              errorMessage = 'There is no user record corresponding to this identifier. The user may have been deleted.';
              break;
            case 'auth/wrong-password':
              errorMessage = 'The password is invalid or the user does not have a password.';
              break;
            case 'auth/invalid-email':
              errorMessage = 'The email address is badly formatted.';
              break;
            default:
              errorMessage = 'Something went wrong. Please try again later.';
          }
      
          Alert.alert(
            "Authentication Failed",
            errorMessage,
            [
              { text: "OK" }
            ]
          );
        }
      },
      signout: async () => {
        try {
          await signOut(auth);
        } catch (e) {
          console.error(e);
        }
      },
    }),
    [user]
  );

  return (
    <AuthContext.Provider value={authContextValue}>
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
