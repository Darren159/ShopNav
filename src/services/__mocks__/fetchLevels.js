const mockFetchLevels = jest.fn();

export default mockFetchLevels;
