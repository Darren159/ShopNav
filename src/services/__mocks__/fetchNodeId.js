const mockFetchNodeId = jest.fn();

export default mockFetchNodeId;
