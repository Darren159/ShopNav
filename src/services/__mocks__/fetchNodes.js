const mockFetchNodes = jest.fn();

export default mockFetchNodes;
