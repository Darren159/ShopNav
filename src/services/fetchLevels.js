import { doc, getDoc } from "firebase/firestore";
import { db } from "../../firebaseConfig";

export default async function fetchLevels(currentMall) {
  if (currentMall) {
    // Check if currentMall is set
    const mallDocRef = doc(db, "malls", currentMall);
    const mallDocSnap = await getDoc(mallDocRef);
    if (mallDocSnap.exists()) {
      const levelsData = Array.from(
        { length: mallDocSnap.data().levels },
        (_, i) => i + 1
      );
      return levelsData;
    }
  }
  throw new Error("Error retrieving level data");
}
