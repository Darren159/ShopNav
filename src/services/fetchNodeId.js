import { doc, getDoc } from "firebase/firestore";
import { db } from "../../firebaseConfig";

// This is an asynchronous function that fetches a specific node ID for a given store in a specific mall from Firebase Firestore.
// Parameters:
// currentMall - The name of the mall where the store is located.
// storeName - The name of the store for which the node ID is to be fetched.
export default async function fetchNodeId(currentMall, storeName) {
  // Format the store name to create a node identifier (lowercase, spaces replaced with dashes, appended with '-node').
  const formattedStoreName = `${storeName
    .replace(/[^\w\s]/g, "")
    .replace(/\s/g, "-")
    .toLowerCase()}-node`;

  // Create a document ID by concatenating the lowercase mall ID and the formatted store name.
  const documentID = `${currentMall.toLowerCase()}-${formattedStoreName}`;

  // Get a reference to the specific document (node) in the Firestore collection.
  const docRef = doc(db, "malls", currentMall, "nodes", documentID);

  // Fetch the document.
  const docSnap = await getDoc(docRef);

  // If the document exists, return the document ID. This is the node ID for the specified store.
  if (docSnap.exists()) {
    return docSnap.id;
  }

  // If the document does not exist (i.e., if the store name is invalid), throw an error.
  throw new Error("Invalid Store Name");
}
