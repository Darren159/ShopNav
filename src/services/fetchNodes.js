import { collection, getDocs } from "firebase/firestore";
import { db } from "../../firebaseConfig";

export default async function fetchNodes(currentMall) {
  const mallNodesCollection = collection(db, "malls", currentMall, "nodes");
  const nodesSnapshot = await getDocs(mallNodesCollection);
  if (nodesSnapshot) {
    const nodeGraph = {};
    nodesSnapshot.docs.forEach((node) => {
      nodeGraph[node.id] = node.data();
    });
    return nodeGraph;
  }
  throw new Error("Error calcualting path");
}
