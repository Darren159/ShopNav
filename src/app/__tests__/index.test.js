import renderer from "react-test-renderer";
import { Alert } from "react-native";
import { fireEvent, waitFor, render } from "@testing-library/react-native";
import Directory from "../index";
import { MallContext } from "../context/mallProvider";
import fetchNodeId from "../../services/fetchNodeId";
import fetchNodes from "../../services/fetchNodes";
import fetchLevels from "../../services/fetchLevels";

jest.mock("firebase/auth", () => ({
  onAuthStateChanged: jest.fn(),
}));

jest.mock("../../../firebaseConfig.js", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../../services/fetchMalls");

jest.mock("../../services/fetchNodeId");

jest.mock("../../services/fetchNodes");

jest.mock("../../services/fetchSVGUrl");

jest.mock("../../services/fetchStoreList");

jest.mock("../../services/fetchNodes.js");

jest.mock("../../services/fetchLevels.js");

const setCurrentMallMock = jest.fn();

const mockedContextValue = {
  currentMall: "Mall 1",
  setCurrentMall: setCurrentMallMock,
  malls: ["Mall 1", "Mall 2", "Mall 3"],
  storeList: [
    { id: "1", name: "Store 1" },
    { id: "2", name: "Store 2" },
  ],
};

fetchLevels.mockResolvedValue([1]);

jest.spyOn(Alert, "alert");

describe("Directory", () => {
  it("renders correctly", async () => {
    const tree = renderer
      .create(
        <MallContext.Provider value={mockedContextValue}>
          <Directory />
        </MallContext.Provider>
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it('renders "Go!" button and shows loading indicator when pressed', async () => {
    fetchNodeId.mockResolvedValue("1");
    fetchNodes.mockResolvedValue({});

    const { getByText, queryByText, getByTestId, queryByTestId } = render(
      <MallContext.Provider value={mockedContextValue}>
        <Directory />
      </MallContext.Provider>
    );

    const goButton = getByText("Go!");
    fireEvent.press(goButton);

    expect(getByTestId("button-loading")).toBeTruthy();
    expect(queryByText("Go!")).toBeNull();

    await waitFor(() => expect(queryByTestId("button-loading")).toBeTruthy());
  });

  it("shows an error alert when fetchNodes fails", async () => {
    const error = new Error("An error occurred");

    fetchNodes.mockRejectedValue(error);

    render(
      <MallContext.Provider value={mockedContextValue}>
        <Directory />
      </MallContext.Provider>
    );

    await waitFor(() =>
      expect(Alert.alert).toHaveBeenCalledWith(
        "Error",
        error.message,
        [{ text: "OK" }],
        { cancelable: false }
      )
    );
  });

  it("shows an error alert when fetchNodeId fails", async () => {
    const error = new Error("An error occurred");

    fetchNodeId.mockRejectedValue(error);

    const { getByText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <Directory />
      </MallContext.Provider>
    );

    const goButton = getByText("Go!");
    fireEvent.press(goButton);

    await waitFor(() =>
      expect(Alert.alert).toHaveBeenCalledWith(
        "Error",
        error.message,
        [{ text: "OK" }],
        { cancelable: false }
      )
    );
  });

  it("shows loading indicator when currentMall is not yet set", async () => {
    const { getByTestId } = render(
      <MallContext.Provider
        value={{ ...mockedContextValue, currentMall: undefined }}
      >
        <Directory />
      </MallContext.Provider>
    );

    expect(getByTestId("map-loading")).toBeTruthy();
  });
});
