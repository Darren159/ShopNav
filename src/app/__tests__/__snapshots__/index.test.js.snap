// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Directory renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 0.15,
        "flexDirection": "row",
        "justifyContent": "center",
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 0.85,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "paddingLeft": 5,
            "width": "100%",
            "zIndex": 5,
          }
        }
      >
        <Text />
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter starting point"
          style={
            {
              "flex": 1,
              "paddingLeft": 5,
              "paddingVertical": 5,
            }
          }
          value=""
        />
      </View>
      <Text />
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "paddingLeft": 5,
            "width": "100%",
            "zIndex": 5,
          }
        }
      >
        <Text />
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter destination"
          style={
            {
              "flex": 1,
              "paddingLeft": 5,
              "paddingVertical": 5,
            }
          }
          value=""
        />
      </View>
    </View>
    <View
      style={
        {
          "width": 75,
        }
      }
    >
      <Text
        style={
          {
            "textAlign": "center",
          }
        }
      >
        Elevators Only
      </Text>
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#81b0ff"
        style={
          {
            "height": 31,
            "width": 51,
          }
        }
        thumbTintColor="#f4f3f4"
        tintColor="#767577"
        value={false}
      />
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "backgroundColor": "#515151",
              "borderRadius": 13,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "padding": 8,
              },
              {
                "backgroundColor": "#515151",
                "borderRadius": 13,
              },
              undefined,
            ]
          }
        >
          <Text />
          <Text
            selectable={false}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "fontWeight": "600",
                },
                {
                  "color": "white",
                },
              ]
            }
          >
            Go!
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 0.85,
      }
    }
  >
    <View
      animatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
              {
                "translateY": 0,
              },
              {
                "scale": 1,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        {
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
            {
              "scale": 1,
            },
          ],
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight="100%"
        bbWidth="100%"
        focusable={false}
        height="100%"
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "flex": 0,
              "height": "100%",
              "width": "100%",
            },
          ]
        }
        vbHeight={760}
        vbWidth={600}
        width="100%"
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
        />
      </RNSVGSvgView>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "bottom": 20,
        "flexDirection": "column-reverse",
        "justifyContent": "space-between",
        "position": "absolute",
        "right": 20,
      }
    }
  />
</RNCSafeAreaView>
`;
